src/lib/comms/sol-lwm2m-server.c

```
@@ -802,34 +812,113 @@ observer_entry_del_monitor(struct observer_entry *entry,
+        return NULL;
+    case SOL_LWM2M_SECURITY_MODE_CERTIFICATE:
+        sec_mode_str = "Certificate";
+        SOL_WRN("Certificate security mode is not supported yet.");
+        return NULL;
+    case SOL_LWM2M_SECURITY_MODE_NO_SEC:
+        sec_mode_str = "NoSec";
+        SOL_DBG("Using NoSec Security Mode (No DTLS).");
+        break;
+    default:
+        SOL_WRN("Unknown DTLS Security Mode: %d", sec_mode);
+        return NULL;
+    }
+
     server = calloc(1, sizeof(struct sol_lwm2m_server));
     SOL_NULL_CHECK(server, NULL);
 
-    server->coap = sol_coap_server_new(&servaddr, false);
+    server->coap = sol_coap_server_new(&servaddr_coap, false);
     SOL_NULL_CHECK_GOTO(server->coap, err_coap);
 
+    if (sec_mode != SOL_LWM2M_SECURITY_MODE_NO_SEC) {
+        if (sec_mode == SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY) {
+            sol_vector_init(&server->known_psks, sizeof(sol_lwm2m_security_psk));
+
+            SOL_VECTOR_FOREACH_IDX (known_psks, given_psk, i) {
+                psk = sol_vector_append(&server->known_psks);
+                SOL_NULL_CHECK_GOTO(psk, err_psk);
+                psk->id = sol_blob_ref(given_psk->id);
+                psk->key = sol_blob_ref(given_psk->key);
+            }
+        }
+
+        server->dtls_server = sol_coap_server_new(&servaddr_dtls, true);
+        SOL_NULL_CHECK_GOTO(server->dtls_server, err_dtls);
+
+        server->security = sol_lwm2m_server_security_add(server, sec_mode);
+        if (!server->security) {
+            if (sec_mode == SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY) {
+                SOL_VECTOR_FOREACH_IDX (&server->known_psks, psk, i) {
+                    sol_blob_unref(psk->id);
+                    sol_blob_unref(psk->key);
+                }
+                sol_vector_clear(&server->known_psks);
+            }
+            sol_coap_server_unref(server->dtls_server);
+            SOL_WRN("Could not enable %s security mode for LWM2M Server", sec_mode_str);
```

@bsmelo You will only create the dtls server is the sec mode is != NO_SEC, thus if you try to create a server_security and it fails, you must not continue. You should return an error!